src/main/java/org/jenkinsci/plugins/workflow/graphanalysis/ForkScanner.java

```diff
@@ -407,27 +382,30 @@ ArrayDeque<ParallelBlockStart> leastCommonAncestor(@NonNull final Set<FlowNode>
                 FlowPiece existingPiece = branches.get(nextBlockStart);
                 if (existingPiece == null && myPiece instanceof FlowSegment) { // No merge, just add to segment
                     ((FlowSegment) myPiece).add(nextBlockStart);
-                    branches.put(nextBlockStart, myPiece);
+                    if (isParallelStart(nextBlockStart)) {
+                        // We make a fork for every parallel step preemptively in case it has less than 2 branches.
+                        Fork f = ((FlowSegment) myPiece).split(branches, (BlockStartNode) nextBlockStart);
+                        // We always create the fork at the end of the existing segment at its end, so we must replace
+                        // the piece with the fork ahead of it.
+                        int headIndex = livePieces.indexOf(myPiece);
+                        livePieces.set(headIndex, f);
+                        parallelForks.add(f);
+                        branches.put(nextBlockStart, f);
+                    } else {
+                        branches.put(nextBlockStart, myPiece);
+                    }
                 } else if (existingPiece == null && myPiece instanceof Fork) {  // No merge, we had a fork. Start a segment preceding the fork
                     FlowSegment newSegment = new FlowSegment();
                     newSegment.isLeaf = false;
                     newSegment.add(nextBlockStart);
-                    newSegment.after = myPiece;
                     pieceIterator.remove();
                     pieceIterator.add(newSegment);
                     branches.put(nextBlockStart, newSegment);
                 } else if (existingPiece != null) {  // Always not null. We're merging into another thing, we're going to eliminate a branch
                     if (existingPiece instanceof Fork) {
                         ((Fork) existingPiece).following.add(myPiece);
-                    } else { // Split a flow segment so it forks against this one
-                        Fork f = ((FlowSegment) existingPiece).split(branches, (BlockStartNode)nextBlockStart, myPiece);
-                        // If we split the existing segment at its end, we created a fork replacing its latest node
-                        // Thus we must replace the piece with the fork ahead of it
-                        if (f.following.contains(existingPiece) ) {
-                            int headIndex = livePieces.indexOf(existingPiece);
-                            livePieces.set(headIndex, f);
-                        }
-                        parallelForks.add(f);
+                    } else {
+                        throw new IllegalStateException("Unexpected join on " + existingPiece);
```

Comment:
Should be unreachable now, since all merges should hit the above condition 🤞

Reply:
Turns out this is still reachable in at least some cases, see `com.cloudbees.workflow.rest.endpoints.ParallelStepTest.test_stages_order` in `pipeline-stage-view`

```
java.lang.IllegalStateException: Unexpected join on org.jenkinsci.plugins.workflow.graphanalysis.ForkScanner$FlowSegment@428b6b34
	at org.jenkinsci.plugins.workflow.graphanalysis.ForkScanner.leastCommonAncestor(ForkScanner.java:414)
	at org.jenkinsci.plugins.workflow.graphanalysis.ForkScanner.setHeads(ForkScanner.java:453)
	at org.jenkinsci.plugins.workflow.graphanalysis.AbstractFlowScanner.setup(AbstractFlowScanner.java:140)
	at org.jenkinsci.plugins.workflow.graphanalysis.AbstractFlowScanner.setup(AbstractFlowScanner.java:151)
	at org.jenkinsci.plugins.workflow.graphanalysis.ForkScanner.visitSimpleChunks(ForkScanner.java:630)
	at com.cloudbees.workflow.rest.external.RunExt.createNew(RunExt.java:323)
	at com.cloudbees.workflow.rest.external.RunExt.create(RunExt.java:311)
	at com.cloudbees.workflow.rest.external.JobExt.create(JobExt.java:143)
	at com.cloudbees.workflow.rest.endpoints.JobAPI.doRuns(JobAPI.java:69)
```

Reply:
I haven't been able to reproduce the failure yet. The test in question starts a build that uses `sleep` in some parallel branches and just checks it as soon as it has a `Run`, so the flow execution is an an unknown and possibly changing state when `ForkScanner` runs.

Reply:
Finally reproduced it, seems to be quite rare (fails about 1 in 40 times for me), so I'm glad it happened to fail in the PCT run!

Reply:
The problematic case seems to be where the parallel step start node is itself a head. That means it gets left out of `blockStartIterator` for all heads, and we end up trying to join two heads on whatever segment has the first parent of the parallel step. It seems to be possible to get things working by restoring some of the old cases in `FlowSegment.split`, but that seems undesirable, because we end up with a `Fork` whose start node is not actually a parallel step, which is bizarre.

Reply:
I have a fix for this, but as I add new test assertions I keep finding new things that don't work and need to compare against the previous behavior to understand if they ever worked, so it is slow going.